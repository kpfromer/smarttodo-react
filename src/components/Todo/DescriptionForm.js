import React, { Component } from 'react';
import PropTypes from 'prop-types';

export default class DescriptionForm extends Component {

  static propTypes = {
    onUpdate: PropTypes.func.isRequired,
    description: PropTypes.string,
    className: PropTypes.string,
    invalidClassName: PropTypes.string,
    clearOnUpdate: PropTypes.bool
  };

  static defaultProps = {
    description: '',
    className: '',
    invalidClassName: '',
    clearOnUpdate: false
  };

  state = {
    description: this.props.description,
    invalid: false
  };

  updateDescription = event => {
    const value = event.target.value;
    this.setState({
      description: value,
      invalid: value.length === 0
    });
  };

  handleSubmit = event => {
    event.preventDefault();
    const { description } = this.state;
    if (description.length > 0) {
      this.props.onUpdate(description);
      if (this.props.clearOnUpdate) {
        this.setState({ description: '' });
      }
    } else {
      this.setState({ invalid: true })
    }
  };

  render() {
    const { description, invalid } = this.state;
    const { className, invalidClassName } = this.props;
    const invalidStyle = invalid ? invalidClassName : '';
    return (
      <form onSubmit={this.handleSubmit}>
        <input
          className={`${className} ${invalidStyle}`}
          value={description}
          onChange={this.updateDescription}
        />
      </form>
    );
  }
}
