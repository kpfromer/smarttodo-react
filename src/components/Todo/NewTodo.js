import React, { PureComponent } from 'react';
import PropTypes from 'prop-types';

import ListItem from '@material-ui/core/ListItem';
import ListItemText from '@material-ui/core/ListItemText';
import DescriptionForm from "./DescriptionForm";

export default class NewTodo extends PureComponent {

  static propTypes = {
    onCreate: PropTypes.func.isRequired
  };

  render() {
    return (
      <ListItem>
        <ListItemText>
          <DescriptionForm clearOnUpdate onUpdate={this.props.onCreate}/>
        </ListItemText>
      </ListItem>
    );
  }
}
