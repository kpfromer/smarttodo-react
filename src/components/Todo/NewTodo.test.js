import { SetupComponent } from "react-component-setup";
import NewTodo from "./NewTodo";
import DescriptionForm from "./DescriptionForm";

const { shallow: setup } = SetupComponent({
  component: NewTodo,
  defaultProps: {
    onCreate: () => {}
  }
});

describe('NewTodo', () => {
  it('renders without crashing', () => {
    setup();
  });

  it('renders a DescriptionForm that clears on update', () => {
    expect(setup().wrapper).toMatchSnapshot();
  });

  it('it triggers onCreate when DescriptionForm updates', () => {
    const value = 'new value';
    const mockHandleCreate = jest.fn();

    const { wrapper } = setup({
      onCreate: mockHandleCreate
    });
    wrapper.find(DescriptionForm).props().onUpdate(value);
    wrapper.update();

    expect(mockHandleCreate).toHaveBeenCalledWith(value);
  });
});