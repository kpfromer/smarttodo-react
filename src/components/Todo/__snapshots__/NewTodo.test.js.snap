// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`NewTodo renders a DescriptionForm that clears on update 1`] = `
<WithStyles(ListItem)>
  <WithStyles(ListItemText)>
    <DescriptionForm
      clearOnUpdate={true}
      description=""
      onUpdate={[Function]}
    />
  </WithStyles(ListItemText)>
</WithStyles(ListItem)>
`;
